source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -49,25 +58,83 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   lookup_ = cache_.makeLookupContext(std::move(lookup_request));
 
   ASSERT(lookup_);
-
   ENVOY_STREAM_LOG(debug, "CacheFilter::decodeHeaders starting lookup", *decoder_callbacks_);
-  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
-  if (state_ == GetHeadersState::GetHeadersResultUnusable) {
+
+  // Used exclusively to access the RequestHeaderMap in onHeaders
+  request_headers_ = &headers;
+  getHeaders();
+
+  if (get_headers_state_ == GetHeadersState::GetHeadersResultUnusable) {
     // onHeaders has already been called, and no usable cache entry was found--continue iteration.
     return Http::FilterHeadersStatus::Continue;
   }
   // onHeaders hasn't been called yet--stop iteration to wait for it, and tell it that we stopped
   // iteration.
-  state_ = GetHeadersState::FinishedGetHeadersCall;
+  get_headers_state_ = GetHeadersState::FinishedGetHeadersCall;
   return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
 Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
                                                      bool end_stream) {
-  // If lookup_ is null, the request wasn't cacheable, so the response isn't either.
-  if (lookup_ && request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers)) {
-    // TODO(yosrym93): Add date internal header or metadata to cached responses and use it instead
-    // of the date header
+  if (skip_encoding_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  filter_state_ = FilterState::Encoding;
+
+  // If lookup_ is null, the request wasn't cacheable, so the response isn't either
+  if (!lookup_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (validating_cache_entry_ && isResponseNotModified(headers)) {
```

Comment:
The body of this `if` seems like a good candidate for a helper, as well as `if (request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers))`.

I'm envisioning something like
```
if (validating_cache_entry_ && isResponseNotModified(headers))
  return processNotModifiedValidation();
else if (request_allows_inserts_ && CacheabilityUtils::isCacheableResponse(headers))
  return processUpdatedValidation();
else
  return Continue;
```

Reply:
I was thinking of this while writing it but I probably forgot when I finished, will try cleaning this part.